_posts/2023-06-19-End-to-end-acknowledgements-in-Data-Prepper.md

```diff
@@ -0,0 +1,64 @@
+
+Data Prepper also provides support for sending negative acknowledgments to indicate explicit failure. The callback function can examine the acknowledgment status, whether it is positive or negative, and take appropriate action based on that information.
+
+In certain scenarios, when Data Prepper sinks are configured with a Dead Letter Queue (DLQ), events that cannot be delivered to the external sink, such as OpenSearch, are written to DLQ. When end-to-end acknowledgments are enabled, successfully writing the events to the DLQ (after failing to deliver them to the external sink) is considered a successful completion of event delivery. In this case, a positive acknowledgment is delivered to the AcknowledgmentSet, indicating successful event processing and handling.
+
+
+## Moving forward: Conclusion and next steps
+
+End-to-end acknowledgments offer robust data durability when using Data Prepper for data processing and ingestion. Some considerations, however, are necessary:
+
+1. End-to-end acknowledgments currently are not compatible with stateful aggregations. The Data Prepper maintainers are considering this issue and exploring solutions for release in future iterations.
+
+2. To leverage end-to-end acknowledgments, source plugins must integrate with the acknowledgment system. Currently, the Amazon S3 source is the only source plugin with this functionality. However, the capability is designed to expand to include additional sources and is well-suited for pull-based sources. Ongoing development of a new Kafka source is showing to be a promising candidate for using these acknowledgments. Users also have expressed interest in applying these acknowledgments to push-based sources like HTTP, which is an area maintainers are exploring.
+
+If you have data stored in Amazon S3 and would like to ingest it into OpenSearch, try out the Data Prepper end-to-end acknowledgments feature. As always, the Data Prepper community appreciates your feedback and contributions.
+
+A brief description of the feature, how to enable it for the sources that currently support the feature, and details on how to add end-to-end acknowledgments to new or other sources can be found [here](https://github.com/opensearch-project/data-prepper/blob/main/docs/end_to_end_acknowledgements.md).
```

Comment:
```suggestion
See [End-to-end Acknowledgments](https://github.com/opensearch-project/data-prepper/blob/main/docs/end_to_end_acknowledgements.md) for more details about the feature, how to enable it for a source, and how to enable it for a sink.
```

Reply:
"how to enable it for a sink" wouldn't be correct. I have replaced it with "what changes needed in a sink to send acknowledgments.".